File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
class Solution {
    public int goodNodes(TreeNode root) {
      int total = 0;
      Stack<Pair> stack = new Stack<>();
      stack.add(new Pair(root, root.val));

      while (!stack.isEmpty()) {
        var cur = stack.pop();

        if (cur.node.val >= cur.maxSoFar) {
          total++;
          if (cur.node.left != null)
            stack.add(new Pair(cur.node.left, cur.node.val));

          if (cur.node.right != null)
            stack.add(new Pair(cur.node.right, cur.node.val));

        } else {
          if (cur.node.left != null)
            stack.add(new Pair(cur.node.left, cur.maxSoFar));

          if (cur.node.right != null)
            stack.add(new Pair(cur.node.right, cur.maxSoFar));
        }
      }

      return total;
    }
}

class Pair {

  TreeNode node;
  int maxSoFar;

  Pair(TreeNode node, int maxSoFar) {
    this.node = node; this.maxSoFar = maxSoFar;
  }

}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cpp
class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        std::stack<TreeNode*> stack;
        stack.push(root);

        while (!stack.empty()) {
            TreeNode* cur = stack.top();
            if (cur == nullptr) {stack.pop(); continue;}

            if (helper(cur, subRoot)) {
                return true;
            }
            stack.pop();
            stack.push(cur->left);
            stack.push(cur->right);
        }

        return false;
    }

    bool helper(TreeNode* node, TreeNode* subRoot) {
        if (node == nullptr && subRoot == nullptr) return true;
        if ((node == nullptr && subRoot != nullptr) ||
            (node != nullptr && subRoot == nullptr)) return false;

        return (
            (node->val == subRoot->val) &&
            helper(node->left, subRoot->left) &&
            helper(node->right, subRoot->right)
        );
    }
};
